runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/DataflowExecutionContext.java

```diff
@@ -341,8 +342,8 @@ public Optional<ActiveMessageMetadata> getActiveMessageMetadata() {
       return Optional.ofNullable(activeMessageMetadata);
     }
 
-    public Map<String, IntSummaryStatistics> getProcessingTimesByStepCopy() {
-      Map<String, IntSummaryStatistics> processingTimesCopy = processingTimesByStep;
+    public synchronized Map<String, IntSummaryStatistics> getProcessingTimesByStepCopy() {
+      Map<String, IntSummaryStatistics> processingTimesCopy = new HashMap<>(processingTimesByStep);
```

Comment:
this is still exposing the IntSummaryStatistics which is not an immutable object.
https://docs.oracle.com/javase/8/docs/api/java/util/IntSummaryStatistics.html

So I think you need to do a deep copy of the values as well.
Map<String, IntSummaryStatistics> processingTimesCopy = processingTimesByStep.entrySet().stream()
    .collect(Collectors.toMap(e -> e.getKey(), e -> e.getValue().clone())

you might need to cast the clone result, not sure.

Reply:
you will also need to synchronize access to processingTimesByStep below

I believe activeMessageMetadata itself is only accessed by the processing thread, so you can just put a synchronized block around the map modification

synchronized (this) {
  this.processingTimesByStep.compute(...);
}

Reply:
the sampler will also access activeMessageMetadata so i think i do need to synchronize that. since getActiveMessageMetadata returns an optional, i dont need to copy it , correct?

as for the deep copy of the IntSummaryStatistics, it didn't like the clone. compiler was throwing `'clone()' has protected access in 'java.lang.Object'`. Would this work?

      Map<String, IntSummaryStatistics> processingTimesCopy =
          processingTimesByStep.entrySet().stream()
              .collect(Collectors.toMap(e -> e.getKey(), e -> {
                IntSummaryStatistics clone = new IntSummaryStatistics();
                clone.combine(e.getValue());
                return clone;
              }));

Reply:
> since getActiveMessageMetadata returns an optional, i dont need to copy it , correct?

We don't need to deep copy there since ActiveMessageMetadata is immutable.

> Would this work?

looks good to me. I would call it copy instead of clone though to not mixup with clone().

Reply:
We also need to guard access to activeMessageMetadata inside enterState()

Reply:
the sampler will also ask to read activeMessageMetadata so i think that needs to be synchronized too.

as for deep copying IntSummaryStatistics values, the compiler doesn't like the clone: `'clone()' has protected access in 'java.lang.Object'`

would something like this work?

      Map<String, IntSummaryStatistics> processingTimesCopy =
          processingTimesByStep.entrySet().stream()
              .collect(Collectors.toMap(e -> e.getKey(), e -> {
                IntSummaryStatistics clone = new IntSummaryStatistics();
                clone.combine(e.getValue());
                return clone;
              }));